reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+                    .flatMap(resolvable -> iterate(estimatesFromResolvable(resolvable)))
+                    .toList();
+            Map<Variable, LocalEstimate> costCover = computeGreedyEstimateCoverForVariables(variableFilter, includedEstimates);
+            long ret = costOfEstimateCover(variableFilter, costCover);
+            assert ret > 0;             // Flag in tests if it happens.
+            return Math.max(ret, 1);    // Don't do stupid stuff in prod when it happens.
+        }
+
+        private Map<Variable, LocalEstimate> computeGreedyEstimateCoverForVariables(Set<Variable> variableFilter, List<LocalEstimate> includedEstimates) {
+            // Does a greedy set cover
+            Map<Variable, LocalEstimate> costCover = new HashMap<>(unaryEstimateCover);
+            includedEstimates.sort(Comparator.comparing(x -> x.answerEstimate(variableFilter)));
+            for (LocalEstimate estimate : includedEstimates) {
+                Set<Variable> filteredVariablesInEstimate = estimate.variables.stream()
+                        .filter(variableFilter::contains).collect(Collectors.toSet());
+
+                long currentCostToCover = costOfEstimateCover(filteredVariablesInEstimate, costCover);
+                if (currentCostToCover > estimate.answerEstimate(filteredVariablesInEstimate)) {
+                    filteredVariablesInEstimate.forEach(v -> costCover.put(v, estimate));
+                }
+            }
+            return costCover;
+        }
+
+        private static long costOfEstimateCover(Set<Variable> variablesToConsider, Map<Variable, LocalEstimate> coverMap) {
+            Set<LocalEstimate> subsetCoveredBy = coverMap.keySet().stream().filter(variablesToConsider::contains)
+                    .map(coverMap::get).collect(Collectors.toSet());
+            return subsetCoveredBy.stream().map(estimate -> estimate.answerEstimate(variablesToConsider)).reduce(1L, (x, y) -> x * y);
+        }
+
+        public boolean mayInitialize() {
+            return initializationStatus == InitializationStatus.NOT_STARTED || initializationStatus == InitializationStatus.RESET;
+        }
+
+        private void reset() {
+            assert this.initializationStatus == InitializationStatus.COMPLETE;
+            this.initializationStatus = InitializationStatus.RESET;
+            this.inferrableEstimates = null;
+            this.fullAnswerCount = -1;
+            this.negatedsCost = -1;
+        }
+
+        private void registerTriggeredRules() {
+            resolvables().filter(this::isInferrable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+                Map<Rule, Set<Unifier>> unifiers = logicMgr.applicableRules(concludable);
+                iterate(unifiers.keySet()).forEachRemaining(rule -> {
+                    answerCountEstimator.registerConjunctionAndInitialize(rule.condition().conjunction());
+                });
+            });
+        }
+
+        private void initialize() {
```

Comment:
if this is `computeEstimate()`, which does the main work of computing the estimate...
Then if we see what `registerTriggeredRules` does, it actually computes the estimates of the recursive rules so it should be called `computeRecursiveEstimates`

We can then follow that naming convention to replace the use of `estimatexxxxInferrables` with using the `recursive` term so we're consistent (general practice - a different term somewhere can really help align the different methods and state everywhere that weren't quite smooth reading before!)